### PR TITLE
Alpha beta

### PR DESCRIPTION
**Pull request by cryos**
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

### CODE REVIEW

**Comment by ghutchis** [comment]
I'll build and review a bit later, but this will be a definite advance over Avo1.x.

File: avogadro/io/cjsonformat.cpp [hunk 1]
```diff
@@ -387,13 +398,23 @@ bool CjsonFormat::write(std::ostream& file, const Molecule& molecule)
...
         moCoefficients.push_back(moMatrix(i, j));
-      }
+
+    if (betaMatrix.cols() > 0 && betaMatrix.rows() > 0) {
+      json moBeta;
+      for (int j = 0; j < moMatrix.cols(); ++j)
+        for (int i = 0; i < moMatrix.rows(); ++i)
+          moBeta.push_back(moMatrix(i, j));
+
+      root["orbitals"]["alpha"] = moCoefficients;
```
Comment (ghutchis) [inline avogadro/io/cjsonformat.cpp]: I think it should be clearer that these are coefficients. The key could also imply energies/eigenvalues.

How about "alphaCoeff" or "alphaCoefficients"

File: avogadro/io/cjsonformat.cpp [hunk 2]
```diff
@@ -257,6 +257,17 @@ bool CjsonFormat::read(std::istream& file, Molecule& molecule)
         for (unsigned int i = 0; i < moCoefficients.size(); ++i)
           coeffs.push_back(static_cast<double>(moCoefficients[i]));
         basis->setMolecularOrbitals(coeffs);
+      } else if (isNumericArray(moCoefficientsA) &&
+                 isNumericArray(moCoefficientsB) &&
+                 moCoefficientsA.size() == moCoefficientsB.size()) {
```
Comment (ghutchis) [inline avogadro/io/cjsonformat.cpp]: I have a nagging feeling that some programs may not keep this last equivalence (e.g., # of alpha and # of beta orbitals may be different)

**Comment by ghutchis** [comment]
My main concern is the choice of json key for alpha/beta MO coefficients. I'd love to see eigenvalue/energies and occupancies come through, but those can be separate patches.

File: avogadro/io/cjsonformat.cpp (hunk 1 above)
Comment (cryos) [reply avogadro/io/cjsonformat.cpp]: Sounds good, this is pretty minor, I can update it.

File: avogadro/io/cjsonformat.cpp (hunk 2 above)
Comment (cryos) [reply avogadro/io/cjsonformat.cpp]: I wondered, I can fix the reader to be a little more relaxed/do two loops...

**Comment by cryos** [comment]
I think this addresses your concerns, absolutely agree about feeding more data through in follow up patches. I really want to get to a point where this format fully round trips, and can help diagnose if other formats are being parsed correctly.

**Pull request merged**